pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +100,255 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
+	// Add a learner member if next peer found
+	var errs []error
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("could not get etcd peerURL to add :%w", err))
+	}
+	if len(peerURL) > 0 {
+		err = c.etcdClient.MemberAddAsLearner(ctx, peerURL)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to add learner member :%w", err))
+		}
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Attempt to promote a learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		errs = append(errs, fmt.Errorf("failed to promote learner: %w", err))
+	}
+
+	return kerrors.NewAggregate(errs)
+}
+
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return "", err
 	}
 
-	if machineForHostPendingDeletion, err := c.isMachinePendingDeletionFor(etcdHost); err != nil {
-		return err
-	} else if machineForHostPendingDeletion {
-		return nil
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return "", err
 	}
 
-	recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s", net.JoinHostPort(etcdHost, "2380")))
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(2).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
+			continue
+		}
+
+		peerURL, err := c.getPeerURLForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get peerURL for node: %v", err)
+		}
+
+		// Ignore if the node's PeerURL is mapped to a member and already part of the quorum.
+		if isURLMappedToMember(peerURL, members) {
+			continue
+		}
+
+		// If the Machine API is unavailable then we can't evaluate the machine's
+		// deletion timestamp or deletion hooks so we scale it up without any considerations
+		// for vertical scaling
+		// This allows UPI clusters without a machine API to scale up
+		// Check machine API availability
+		isMachineAPIFunctional, err := c.machineAPIChecker.IsFunctional()
+		if err != nil {
+			return "", fmt.Errorf("failed to determine Machine API availability: %v", err)
+		}
+		if !isMachineAPIFunctional {
+			return peerURL, nil
+		}
+
+		// If the Machine API is functional we should ignore the machine
+		// if it is pending deletion to avoid re-adding previously removed members
+		internalIP, err := c.getInternalIPForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get internal IP for node: %v", err)
```

Comment:
do we have to stop here? we could collect the errors inside this method (`getEtcdPeerURLToAdd`) and go on to the next node. I think the order doesn't really matter, right?